### PR TITLE
Provide support for commands

### LINKED ISSUE #11: Fix triggering code action from the horizontal ruler
Am I correct that `textDocument/codeActions` requests aren't happening? I'm guessing this is a Monaco problem or...?

### LINKED ISSUE #12: Services (Commands, Workspace) implementations for the standalone editor
Ignoring the fact that `textDocument/codeActions` doesn't seem to be working for me (#11), I can't find any references to the `workspace/executeCommand` request in the code. Is this supported?

### PR DESCRIPTION
**Pull request by gatesn**
This implements a basic applyEdit operation (more advanced ones would accept the editor and preserve / manipulate the current selections), as well as registering commands against a given editor.

Fixes #11 
Fixes #12

### CODE REVIEW

File: src/workspace.ts [hunk 1]
```diff
@@ -83,4 +83,36 @@ export class MonacoWorkspace implements Workspace {
         return this.onDidChangeTextDocumentEmitter.event;
     }
 
+    public applyEdit(workspaceEdit: WorkspaceEdit): Thenable<boolean> {
+        let applied = true;
+        if (workspaceEdit.documentChanges) {
```
Comment (akosyakov) [inline src/workspace.ts]: please use [asWorkspaceEdit](https://github.com/TypeFox/monaco-languageclient/blob/master/src/converter.ts#L273), applying `documentChanges` is not enough

File: src/workspace.ts [hunk 2]
```diff
@@ -83,4 +83,36 @@ export class MonacoWorkspace implements Workspace {
         return this.onDidChangeTextDocumentEmitter.event;
     }
 
+    public applyEdit(workspaceEdit: WorkspaceEdit): Thenable<boolean> {
```
Comment (akosyakov) [inline src/workspace.ts]: please return `Promise`

**Comment by gatesn** [comment]
Actually, having looked at that disgusting applyEdit function, we should iterate over the changes and then apply them all in bulk. Otherwise the undo stack looks a bit odd.

File: src/workspace.ts [hunk 3]
```diff
@@ -83,4 +83,36 @@ export class MonacoWorkspace implements Workspace {
...
+        let applied = true;
+        if (workspaceEdit.documentChanges) {
+            for (const change of workspaceEdit.documentChanges) {
+                if (change.textDocument.version && change.textDocument.version >= 0) {
```
Comment (akosyakov) [inline src/workspace.ts]: Are these checks necessary?

File: src/workspace.ts [hunk 4]
```diff
@@ -83,4 +83,36 @@ export class MonacoWorkspace implements Workspace {
...
+                                    forceMoveMarkers: true,
+                                };
+                            }),
+                            () => [],  // Do not try and preserve editor selections.
+                        );
+                    } else {
+                        applied = false;
+                    }
+                } else {
+                    applied = false;
```
Comment (akosyakov) [inline src/workspace.ts]: Is not the whole change should be applied or rejected? It seems that changes can be applied partially by this method.

File: src/workspace.ts [hunk 5]
```diff
@@ -83,4 +83,36 @@ export class MonacoWorkspace implements Workspace {
...
     }
 
+    public applyEdit(workspaceEdit: WorkspaceEdit): Thenable<boolean> {
+        let applied = true;
+        if (workspaceEdit.documentChanges) {
+            for (const change of workspaceEdit.documentChanges) {
+                if (change.textDocument.version && change.textDocument.version >= 0) {
+                    const textDocument = this.documents.get(change.textDocument.uri);
+                    if (textDocument && textDocument.version === change.textDocument.version) {
+                        monaco.editor.getModel(monaco.Uri.parse(textDocument.uri)).pushEditOperations(
```
Comment (akosyakov) [inline src/workspace.ts]: Should not you check that the model is defined?

File: src/commands.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) 2017 TypeFox GmbH (http://www.typefox.io). All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+import { Commands, Disposable } from 'vscode-base-languageclient/lib/services';
+
+export class MonacoCommands implements Commands {
+
+    public constructor(private _editor: monaco.editor.IStandaloneCodeEditor) { }
```
Comment (akosyakov) [inline src/commands.ts]: `protected readonly editor`

File: src/commands.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+    public registerCommand(command: string, callback: (...args: any[]) => any, thisArg?: any): Disposable {
+        return (this._editor as any)._commandService.addCommand(command, {
```
Comment (akosyakov) [inline src/commands.ts]: please introduce augmenting typings for internals instead of using any

File: src/commands.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+            handler: (id: string, ...args: any[]) => {
+                console.log("Executing command", command, id, args);
```
Comment (akosyakov) [inline src/commands.ts]: Won't it pollute console?

File: src/workspace.ts [hunk 9]
```diff
@@ -83,4 +83,36 @@ export class MonacoWorkspace implements Workspace {
...
+                        applied = false;
+                    }
+                } else {
+                    applied = false;
+                }
+            }
+        } else {
+            applied = false;
+        }
+        return Promise.resolve(applied);
```
Comment (akosyakov) [inline src/workspace.ts]: Could you use high-level functions as `reduce` to get rid of if-else cascade?

File: src/commands.ts (hunk 8 above)
Comment (gatesn) [inline src/commands.ts]: Sorry, left in from testing!

File: src/commands.ts (hunk 7 above)
Comment (gatesn) [inline src/commands.ts]: Could you explain what you mean here? Not sure I'm familiar with this technique.

Comment (akosyakov) [inline src/commands.ts]: I meant that you should add d.ts. file that augments `monaco` module add exposes `_commandService`, look at https://github.com/theia-ide/theia/blob/master/packages/monaco/src/typings/monaco/index.d.ts.

**Comment by gatesn** [comment]
I've made the changes but not written an example. Any ideas on what to use as an example?!

**Comment by akosyakov** [comment]
@gatesn the same here please use rebase instead of merge

**Comment by akosyakov** [comment]
I like changes but there should be a proof that it is working: tests or use of it in the example.

What if the JSON server provides:
- a command that converts a property name at the position to the lower via `applyEdit`
- some way to trigger it for example via the code action

It should not be sophisticated, but only demonstrate that everything works together.

**Comment by gatesn** [comment]
@akosyakov I created an example that provides a code action for any document in all positions. This action triggers a server-side command that upper cases the entire document.

**Comment by akosyakov** [comment]
@gatesn I've tested this PR and it works nicely, thank you! I've opened a new PR against your, please merge it and after that, I will approve this PR.

Next time you can work in the branch directly in this repo since you have the write access instead of working on the fork. It makes easy to push new commits to PRs.

**Comment by gatesn** [comment]
@akosyakov merged your changes, thanks

**Comment by gatesn** [comment]
Do you think we could cut a release after this merges?

**Comment by akosyakov** [comment]
@gatesn LGTM, thank you again!

**Comment by akosyakov** [comment]
> Do you think we could cut a release after this merges?

yes, we can make a release. Should we migrate to [monaco 0.10.0](https://github.com/TypeFox/monaco-languageclient/issues/26) before the release?

**Comment by akosyakov** [comment]
btw i wonder why github cannot recognize your account as an author, see: https://github.com/TypeFox/monaco-languageclient/commits/master

**Comment by gatesn** [comment]
Probably no 0.10.0 given the breaking changes around Code actions. We might need to investigate to see what that means.

**Comment by gatesn** [comment]
And it seems my git setup is different on different laptop :/

**Comment by akosyakov** [comment]
@gatesn 0.2.0 is published with 0.10.0 Monaco support thanks to @rcjsuen

**Pull request merged**